__tests__/session.spec.ts

```diff
@@ -0,0 +1,647 @@
+        request,
+        true,
+        {
+          enabled: false,
+          unauthenticatedPaths: [],
+        },
+        process.env.NEXT_PUBLIC_WORKOS_REDIRECT_URI as string,
+        [],
+      );
+
+      expect(result.status).toBe(200);
+      expect(console.log).toHaveBeenCalledWith(
+        `Session invalid. Refreshing access token that ends in ${mockSession.accessToken.slice(-10)}`,
+      );
+      expect(console.log).toHaveBeenCalledWith(expect.stringContaining('Refresh successful. New access token ends in'));
+    });
+
+    it('should delete the cookie and redirect when refreshing fails', async () => {
+      jest.spyOn(console, 'log').mockImplementation(() => {});
+
+      mockSession.accessToken = await generateTestToken({}, true);
+
+      const nextCookies = await cookies();
+      nextCookies.set(
+        'wos-session',
+        await sealData(mockSession, { password: process.env.WORKOS_COOKIE_PASSWORD as string }),
+      );
+
+      (jwtVerify as jest.Mock).mockImplementation(() => {
+        throw new Error('Invalid token');
+      });
+
+      jest
+        .spyOn(workos.userManagement, 'authenticateWithRefreshToken')
+        .mockRejectedValue(new Error('Failed to refresh'));
+
+      const request = new NextRequest(new URL('http://example.com'));
+
+      const result = await updateSession(
+        request,
+        true,
+        {
+          enabled: false,
+          unauthenticatedPaths: [],
+        },
+        process.env.NEXT_PUBLIC_WORKOS_REDIRECT_URI as string,
+        [],
+      );
+
+      expect(result.status).toBe(307);
+      expect(nextCookies.get('wos-session')).toBeUndefined();
+      expect(console.log).toHaveBeenCalledTimes(2);
+      expect(console.log).toHaveBeenNthCalledWith(
+        1,
+        `Session invalid. Refreshing access token that ends in ${mockSession.accessToken.slice(-10)}`,
+      );
+      expect(console.log).toHaveBeenNthCalledWith(
+        2,
+        'Failed to refresh. Deleting cookie and redirecting.',
+        new Error('Failed to refresh'),
+      );
+    });
+
+    describe('middleware auth', () => {
+      it('should redirect unauthenticated users on protected routes', async () => {
+        jest.spyOn(console, 'log').mockImplementation(() => {});
+
+        const request = new NextRequest(new URL('http://example.com/protected'));
+        const result = await updateSession(
+          request,
+          true,
+          {
+            enabled: true,
+            unauthenticatedPaths: [],
+          },
+          process.env.NEXT_PUBLIC_WORKOS_REDIRECT_URI as string,
+          [],
+        );
+
+        expect(result.status).toBe(307);
+        expect(console.log).toHaveBeenCalledWith(
+          'Unauthenticated user on protected route http://example.com/protected, redirecting to AuthKit',
+        );
+      });
+
+      it('should use Response if NextResponse.redirect is not available', async () => {
+        const originalRedirect = NextResponse.redirect;
```

Comment:
another place we could probably use spyOn ?

Reply:
In this case we actually do want `redirect` to be undefined since we're simulating the conditions of <= Next.js 13.